src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php

```diff
@@ -27,11 +33,60 @@ class EntityType extends DoctrineType
      */
     public function getLoader(ObjectManager $manager, $queryBuilder, $class)
     {
-        return new ORMQueryBuilderLoader(
-            $queryBuilder,
-            $manager,
-            $class
-        );
+        $queryBuilderHash = null;
+
+        if ($queryBuilder instanceof QueryBuilder) {
+            $queryBuilderHash = $this->getQueryBuilderHash($queryBuilder);
+        }
+
+        if (null === $queryBuilderHash) {
+            return new ORMQueryBuilderLoader(
+                $queryBuilder,
+                $manager,
+                $class
+            );
+        }
+
+        $loaderHash = $this->getLoaderHash($manager, $queryBuilderHash, $class);
+
+        if (!isset($this->loaderCache[$loaderHash])) {
+            $this->loaderCache[$loaderHash] = new ORMQueryBuilderLoader(
+                $queryBuilder,
+                $manager,
+                $class
+            );
+        }
+
+        return $this->loaderCache[$loaderHash];
+    }
+
+    /**
+     * @param QueryBuilder $queryBuilder
+     *
+     * @return string
+     */
+    protected function getQueryBuilderHash(QueryBuilder $queryBuilder)
+    {
+        return hash('sha256', json_encode(array(
```

Comment:
@timglabisch well i don't like it, but i made it as in DoctrineType: https://github.com/symfony/symfony/blob/2.7/src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php#L117